util/topsql/reporter/subscriber.go

```diff
@@ -0,0 +1,312 @@
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPublisher implements TopSQLPublisher.
+//
+// If a client subscribes to TopSQL records, the TopSQLPublisher is responsible for registering them to the reporter.
+// Then the reporter sends data to the client periodically.
+type TopSQLPublisher struct {
+	decodePlan     planBinaryDecodeFunc
+	clientRegistry *ReportClientRegistry
+}
+
+// NewTopSQLPublisher creates a new TopSQLPublisher.
+func NewTopSQLPublisher(
+	decodePlan planBinaryDecodeFunc,
+	clientRegistry *ReportClientRegistry,
+) *TopSQLPublisher {
+	return &TopSQLPublisher{
+		decodePlan:     decodePlan,
+		clientRegistry: clientRegistry,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPublisher{}
+
+// Subscribe registers clients to the reporter and redirects data received from reporter
+// to subscribers associated with those clients.
+func (t *TopSQLPublisher) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	sc := newSubClient(stream, t.decodePlan)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go sc.run(&wg)
+
+	t.clientRegistry.register(sc)
+
+	wg.Wait()
+	return nil
+}
+
+type subClient struct {
+	stream   tipb.TopSQLPubSub_SubscribeServer
+	sendTask chan struct {
+		data    reportData
+		timeout time.Duration
+	}
+	isDown *atomic.Bool
+
+	decodePlan planBinaryDecodeFunc
+}
+
+func newSubClient(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+	decodePlan planBinaryDecodeFunc,
+) *subClient {
+	sendTask := make(chan struct {
+		data    reportData
+		timeout time.Duration
+	})
+	return &subClient{
+		stream:   stream,
+		sendTask: sendTask,
+		isDown:   atomic.NewBool(false),
+
+		decodePlan: decodePlan,
+	}
+}
+
+func (s *subClient) run(wg *sync.WaitGroup) {
+	defer func() {
+		wg.Done()
+		s.isDown.Store(true)
+	}()
+
+	for task := range s.sendTask {
+		ctx, cancel := context.WithTimeout(context.Background(), task.timeout)
+		start := time.Now()
+
+		var err error
+		doneCh := make(chan struct{})
+		go func(
+			task struct {
+				data    reportData
+				timeout time.Duration
+			}) {
+			util.WithRecovery(func() {
+				defer func() {
+					doneCh <- struct{}{}
+				}()
+				err = s.doSend(ctx, task.data)
+				if err != nil {
+					reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
+				} else {
+					reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
+				}
+			}, nil)
+		}(task)
+
+		select {
+		case <-doneCh:
+			cancel()
+			if err != nil {
+				logutil.BgLogger().Warn(
+					"[top-sql] client failed to send data to subscriber",
+					zap.Error(err),
+				)
+				return
+			}
+		case <-time.After(task.timeout):
+			cancel()
+			logutil.BgLogger().Warn(
+				"[top-sql] client failed to send data to subscriber due to timeout",
+				zap.Duration("timeout", task.timeout),
+			)
+			return
+		}
+	}
+}
+
+func (s *subClient) doSend(ctx context.Context, data reportData) error {
```

Comment:
```suggestion
	for task := range s.sendTask {
		start := time.Now()
		err := s.doSend(task.data, task.timeout)
		if err != nil {
			reportAllDurationFailedHistogram.Observe(time.Since(start).Seconds())
			logutil.BgLogger().Warn(
				"[top-sql] client failed to send data to subscriber",
				zap.Error(err),
			)
		} else {
			reportAllDurationSuccHistogram.Observe(time.Since(start).Seconds())
		}
	}
}

func (s *subClient) doSend(data reportData, timeout time.Duration) error {
	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()
```

Reply:
Sending through grpc stream doesn't take `ctx` as an argument so it won't respect the timeout. I have to handle timeout manually.

Reply:
em... a little bit complicated...
if the `stream` client is timeout, even you handle timeout manually this time, but you will still meet this problem next time.